In [22]:
I = [[1,9,8,7],
     [1,1,2,8],
     [2,0,1,7],
     [0,4,2,8]]

In [5]:
# TODO 返回矩阵的行数和列数
def shape(M):
    r = len(M)
    c = len(M[0])
    return r,c

In [6]:
# 运行以下代码测试你的 shape 函数
%run -i -e test.py LinearRegressionTestCase.test_shape

.
----------------------------------------------------------------------
Ran 1 test in 0.001s

OK


In [7]:
# TODO 每个元素四舍五入到特定小数数位
# 直接修改参数矩阵，无返回值
def matxRound(M, decPts=4):
    for p in M:
        for index in range(len(M[0])):
            p[index] = round(p[index], decPts)
    pass

In [8]:
# 运行以下代码测试你的 matxRound 函数
%run -i -e test.py LinearRegressionTestCase.test_matxRound

.
----------------------------------------------------------------------
Ran 1 test in 0.059s

OK


In [9]:
# TODO 计算矩阵的转置
def transpose(M):
    trans_M = zip(*M)
    return trans_M

In [10]:
# 运行以下代码测试你的 transpose 函数
%run -i -e test.py LinearRegressionTestCase.test_transpose

.
----------------------------------------------------------------------
Ran 1 test in 0.006s

OK


In [23]:
# TODO 计算矩阵乘法 AB，如果无法相乘则raise ValueError
def matxMultiply(A, B):
    r_A = len(A)
    c_A = len(A[0])
    r_B = len(B)
    c_B = len(B[0])
    multi_C = [[0]*c_B for i in range(r_A)]
    if c_A != r_B:
        raise ValueError("A can't mutiply with B")
    else:
        for i in range(r_A):
            for j in range(c_B):
                for k in range(c_A):
                    multi_C[i][j] += A[i][k]*B[k][j]
    return multi_C

In [24]:
# 运行以下代码测试你的 matxMultiply 函数
%run -i -e test.py LinearRegressionTestCase.test_matxMultiply

.
----------------------------------------------------------------------
Ran 1 test in 0.047s

OK


In [3]:
# TODO 构造增广矩阵，假设A，b行数相同
def augmentMatrix(A, b):
    result = [[0]*(len(A[0])+1) for i in range(len(A))]
    for i in range(len(A)):
        for j in range(len(A[0])+1):
            if j < len(A[0]):
                result[i][j] = A[i][j]
            else:
                result[i][j] = b[i][0]
    return result

In [4]:
# 运行以下代码测试你的 augmentMatrix 函数
%run -i -e test.py LinearRegressionTestCase.test_augmentMatrix

.
----------------------------------------------------------------------
Ran 1 test in 0.006s

OK


In [5]:
# TODO r1 <---> r2
# 直接修改参数矩阵，无返回值
def swapRows(M, r1, r2):
    M[r1], M[r2] = M[r2], M[r1]

In [6]:
# 运行以下代码测试你的 swapRows 函数
%run -i -e test.py LinearRegressionTestCase.test_swapRows

.
----------------------------------------------------------------------
Ran 1 test in 0.002s

OK


In [5]:
# TODO r1 <--- r1 * scale
# scale为0是非法输入，要求 raise ValueError
# 直接修改参数矩阵，无返回值
def scaleRow(M, r, scale):
    if scale != 0:
        for i in range(len(M[r])):
            M[r][i] = M[r][i]*scale
    else:
        raise ValueError("scale is 0")

In [6]:

# 运行以下代码测试你的 scaleRow 函数
%run -i -e test.py LinearRegressionTestCase.test_scaleRow

.
----------------------------------------------------------------------
Ran 1 test in 0.002s

OK


In [7]:
# TODO r1 <--- r1 + r2*scale
# 直接修改参数矩阵，无返回值
def addScaledRow(M, r1, r2, scale):
    if scale != 0:
        for i in range(len(M[r1])):
            M[r1][i] += M[r2][i]*scale
    else:
        raise ValueError("scale is 0")

In [8]:
# 运行以下代码测试你的 addScaledRow 函数
%run -i -e test.py LinearRegressionTestCase.test_addScaledRow

.
----------------------------------------------------------------------
Ran 1 test in 0.004s

OK


In [27]:
# 不要修改这里！
from helper import *
from fractions import Fraction
A = generateMatrix(3,5,singular=False)
b = np.ones(shape=(3,1),dtype=int) # it doesn't matter
Ab = augmentMatrix(A.tolist(),b.tolist()) # 请确保你的增广矩阵已经写好了
printInMatrixFormat(Ab,padding=3,truncating=0)

def swapRows(M, r1, r2):
    M[r1], M[r2] = M[r2], M[r1]
    
def scaleRow(M, r, scale):
    if scale != 0:
        for i in range(len(M[r])):
            M[r][i] = M[r][i]*scale
    else:
        raise ValueError("scale is 0")

def addScaledRow(M, r1, r2, scale):
    if scale != 0:
        for i in range(len(M[r1])):
            M[r1][i] += M[r2][i]*scale
    else:
        raise ValueError("scale is 0")

def gj_Solve(A, b, decPts, epsilon):
    if len(A) != len(b):    #如果 A，b 高度不同，返回None
        return None
    M = augmentMatrix(A.tolist(),b.tolist())
    #print M
    for i in range(len(M[0])-1):    #取第i列
        #print 'i:',i
        value_list = [0]*(len(M)-i)
        for j in range(i,len(M)):
            value = M[j][i]
            value_list[j-i] = abs(value)
        #print 'value_list:',value_list    #对角线及对角线以下值绝对值的列表
        max_value = max(value_list)    #取第i列对角线及以下数值绝对值的最大值
        #print 'max_value:',max_value
        if max_value == epsilon:    #如果最大值为0，则A为奇异矩阵，返回None
            return None
        else:
            for r in range(len(value_list)):
                if value_list[r] == max_value:    #取绝对值最大值所在行
                    #print '最大值所在行:',M[r+i]
                    #print '第i行:',M[i]
                    swapRows(M, i, r+i)    #绝对值最大值所在行与i行交换
                    #print '交换后的第i行:',M[i]
                    scaleRow(M, i, Fraction(1, M[i][i]))     #1.0/M[i][i] 第i行对角线值变为1
                    #print '对角线值为1的第i行',M[i]
                    for l in range(len(M)):    #第i列除对角线值外的元素变为0
                        if l != i:
                            #print '第l行:',l
                            #print "第l行",M[l]
                            addScaledRow(M, l, i, -M[l][i])    #第i列第l个值变为0
                            #print '第i列第l个值为0的l行:',M[l]
                        else:
                            continue
                    print M    #每一列循环后的M
                    break    #退出r的循环
gj_Solve(A, b, decPts=4, epsilon = 1.0e-16)

 -7,  4,  5 ||  1 
 -4,  6, -1 ||  1 
 -2, -6, -3 ||  1 
[[Fraction(1, 1), Fraction(-4, 7), Fraction(-5, 7), Fraction(-1, 7)], [Fraction(0, 1), Fraction(26, 7), Fraction(-27, 7), Fraction(3, 7)], [Fraction(0, 1), Fraction(-50, 7), Fraction(-31, 7), Fraction(5, 7)]]
[[Fraction(1, 1), Fraction(0, 1), Fraction(-9, 25), Fraction(-1, 5)], [Fraction(0, 1), Fraction(1, 1), Fraction(31, 50), Fraction(-1, 10)], [Fraction(0, 1), Fraction(0, 1), Fraction(-154, 25), Fraction(4, 5)]]
[[Fraction(1, 1), Fraction(0, 1), Fraction(0, 1), Fraction(-19, 77)], [Fraction(0, 1), Fraction(1, 1), Fraction(0, 1), Fraction(-3, 154)], [Fraction(0, 1), Fraction(0, 1), Fraction(1, 1), Fraction(-10, 77)]]


$Ab = \left\lgroup \matrix{-7 & 4 & 5 & 1\cr -4 & 6 & -1&1\cr -2&-6&-3&1\cr} \right\rgroup\\
-->\left\lgroup \matrix{1 & \frac{-4}{7} & \frac{-5}{7} & \frac{-1}{7}\cr 0 & \frac{26}{7} & \frac{-27}{7}&\frac{3}{7}\cr 0&\frac{-50}{7}&\frac{-31}{7}&\frac{5}{7}\cr} \right\rgroup\\
-->\left\lgroup \matrix{1 & 0 & \frac{-9}{25} & \frac{-1}{5}\cr 0 & 1 & \frac{31}{50}&\frac{-1}{10}\cr 0&0&\frac{-154}{25}&\frac{4}{5}\cr} \right\rgroup\\
-->\left\lgroup \matrix{1 & 0 & 0 & \frac{-19}{77}\cr 0 & 1 & 0&\frac{-3}{154}\cr 0&0&1&\frac{-10}{77}\cr} \right\rgroup
$


In [41]:
# TODO 实现 Gaussain Jordan 方法求解 Ax = b

""" Gaussian Jordan 方法求解 Ax = b.
    参数
        A: 方阵 
        b: 列向量
        decPts: 四舍五入位数，默认为4
        epsilon: 判读是否为0的阈值，默认 1.0e-16
        
    返回列向量 x 使得 Ax = b 
    返回None，如果 A，b 高度不同
    返回None，如果 A 为奇异矩阵
"""

def gj_Solve(A, b, decPts=4, epsilon=1.0e-16):
    if len(A) != len(b):    #如果 A，b 高度不同，返回None
        return None
    M = augmentMatrix(A.tolist(),b.tolist())
    print M
    for i in range(len(M[0])-1):    #取第i列
        #print 'i:',i
        value_list = [0]*(len(M)-i)
        for j in range(i,len(M)):
            value = M[j][i]
            value_list[j-i] = abs(value)
        #print 'value_list:',value_list    #对角线及对角线以下值绝对值的列表
        max_value = max(value_list)    #取第i列对角线及以下数值绝对值的最大值
        #print 'max_value:',max_value
        if max_value == epsilon:    #如果最大值为0，则A为奇异矩阵，返回None
            return None
        else:
            for r in range(len(value_list)):
                if value_list[r] == max_value:    #取绝对值最大值所在行
                    #print '最大值所在行:',M[r+i]
                    #print '第i行:',M[i]
                    swapRows(M, i, r+i)    #绝对值最大值所在行与i行交换
                    #print '交换后的第i行:',M[i]
                    scaleRow(M, i, Fraction(1, M[i][i]))     #1.0/M[i][i] 第i行对角线值变为1
                    #print '对角线值为1的第i行',M[i]
                    for l in range(len(M)):    #第i列除对角线值外的元素变为0
                        if l != i:
                            #print '第l行:',l
                            #print "第l行",M[l]
                            addScaledRow(M, l, i, -M[l][i])    #第i列第l个值变为0
                            #print '第i列第l个值为0的l行:',M[l]
                        else:
                            continue
                    print M    #每一列循环后的M
                    break    #退出r的循环
    x = [0]*len(b)
    for i in range(len(M)):
        x[i] = M[i][len(M[0])-1]
        
    return x

In [42]:
# 运行以下代码测试你的 gj_Solve 函数
%run -i -e test.py LinearRegressionTestCase.test_gj_Solve

UnicodeDecodeError: 'ascii' codec can't decode byte 0xe7 in position 44: ordinal not in range(128)